### DISCUSSION TITLE
Logs report connection refused for kubernetes.default.svc.cluster.local

### DISCUSSION THREAD

**Discussion by djfinnoy** [kiali/kiali#5971] [Q&A]
Fresh installation of Kiali and I'm seeing some strange error messages in the logs

```
W0331 11:06:29.037920       1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1.ConfigMap: Get "https://10.68.0.1:443/api/v1/configmaps?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

E0331 11:06:29.038089       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Get "https://10.68.0.1:443/api/v1/configmaps?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

W0331 11:06:29.832205       1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1beta1.Gateway: Get "https://10.68.0.1:443/apis/networking.istio.io/v1beta1/gateways?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

E0331 11:06:29.832288       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1beta1.Gateway: failed to list *v1beta1.Gateway: Get "https://10.68.0.1:443/apis/networking.istio.io/v1beta1/gateways?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

W0331 11:06:30.497014       1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1.Pod: Get "https://10.68.0.1:443/api/v1/pods?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

E0331 11:06:30.497090       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: Get "https://10.68.0.1:443/api/v1/pods?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

W0331 11:06:30.576813       1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1.StatefulSet: Get "https://10.68.0.1:443/apis/apps/v1/statefulsets?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

E0331 11:06:30.576934       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.StatefulSet: failed to list *v1.StatefulSet: Get "https://10.68.0.1:443/apis/apps/v1/statefulsets?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

W0331 11:06:30.960861       1 reflector.go:324] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1beta1.AuthorizationPolicy: Get "https://10.68.0.1:443/apis/security.istio.io/v1beta1/authorizationpolicies?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused

E0331 11:06:30.960950       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1beta1.AuthorizationPolicy: failed to list *v1beta1.AuthorizationPolicy: Get "https://10.68.0.1:443/apis/security.istio.io/v1beta1/authorizationpolicies?limit=500&resourceVersion=0": dial tcp 10.68.0.1:443: connect: connection refused
```

This is a Google Kubernetes Engine cluster.
I have installed via the Kiali Operator, and I have installed Istio via the official Helm charts. The only peculiarity is that Kiali is installed in the `istio-system` namespace, and sidecars are enabled for this namespace.
There are no network policies that should affect either `istio-system` or `default`.

**Comment by jmazzitelli** [comment]
Those don't look to be 403s... those look like it just can't connect to the endpoint. `connect: connection refused`

**Reply by djfinnoy** [reply]
Oh yeah, I think my brain read 403 instead of the port number. Question still stands tho, why is Kiali getting connection refused? Are there missing permissions in the service account role?

**Reply by jmazzitelli** [reply]
I can't say I have seen that specific error before in this context - `dial tcp 10.68.0.1:443: connect: connection refused` - that just tells me the Kiali server can't even successfully connect to that endpoint.

I don't think that has anything to do with permissions/service accounts/roles. That error message indicates it is just an outright connection failure - the endpoint isn't there or there is a network problem of some sort. For some reason, the Kiali pod cannot make calls to the k8s API server. Is there any kind of extra security in place within your cluster that prohibits network access to the k8s API? 

You didn't indicate what version of Kiali, Istio, and Kubernetes you are using, or how you installed Kiali (helm chart? operator? Istio addon yaml?). That might be relevant, I don't know. But you should always indicate the versions you are using just so people can know what we are dealing with. And take a look at the Kiali pod logs for any other messages that might be relevant.  Posting the logs could also help people understand what's going on under the covers.

**Reply by djfinnoy** [reply]
Yeah you're right, my apologies. I have updated the original post.

Will look into this more next week. Right now I suspect that maybe there is some sort of Google Kubernetes Engine thing where workloads must have appropriate IAM permissions to communicate with the default service.
